File: learntoprogram/ObjectDemo.java
/**
* Description:  object类常用方法
* @author fnfly2005
* @version 1.0
*/
class PersonObject
{
    private int age;

    PersonObject(int age)
    {
        this.age = age;
    }

    //equals方法默认比较对象间内存地址，覆盖该方法，将其替换成比较内容
    public boolean equals(Object obj)
    {
        if (!(obj instanceof PersonObject))//!() 布尔取反语句,类型判断
        {
            throw new ClassCastException("类型错误");
        }
        PersonObject p = (PersonObject)obj;//向下转型
        return this.age == p.age;
    }

    //toString方法默认返回类名@16进制哈希值，覆盖该方法
    public String toString()
    {
        return "PersonObject:"+age;
    }
}

class ObjectDemo
{
    public static void main (String[] agrs)
    {
        PersonObject p1 = new PersonObject(20);
        PersonObject p2 = new PersonObject(20);

        //equals方法-比较对象
        Object o = new Object();
        System.out.println(p1.equals(p2));
        
        //hashCode方法-返回10进制哈希值-内存地址
        System.out.println(p1);
        System.out.println(Integer.toHexString(p1.hashCode()));//转16进制

        //getClass方法-获取类对象
        Class clazz1 = p1.getClass();
        Class clazz2 = p1.getClass();
        System.out.println(clazz1.getName());//获取类名
        System.out.println(clazz1.equals(clazz2));//不同对象拥有同一个类

        //toString方法-返回对象的字符串表示形式
        System.out.println(p1.toString());
    }
}

File: learntoprogram/ObjectDemo.py
#!/usr/bin/python
#coding:utf-8
##################################
"""
Description: 魔术方法
__str__ __repr__ __cmp__ id __class__.__name__
Version: v1.0
"""
##################################
class PersionObject(object):
    def __init__(self,name,age):
        self.name=name
        self.age=age

    '''
    __str__对应于java中的toString方法,面向用户-print后显示值
    __repr__面向程序员-实际返回值
    id函数对应于java中的hashcode方法，返回对象标识的内存地址，a=b等价于id(a)=id(b)
    '''
    def __str__(self): 
        return "name:" + self.name + "   age:"+str(self.age) + " in " + str(hex(id(self)))
    __repr__=__str__

    '''
    __cmp__对应于java中的equals
    重写cmp函数实现自定义属性比较
    '''
    def __cmp__(self,s):
        if self.age > s.age:
            return 1
        elif self.age < s.age:
            return -1
        else: 
            return 0

if __name__ == '__main__':
    p1=PersionObject('fnfly2005',20)
    p2=PersionObject('fnfly2003',21)
    p3=PersionObject('fnfly2006',31)

    L = [p1,p2,p3]
    print str(sorted(L))+"##cmp比较函数适用于sorted"
    print str(p1 < p2) +"##自定义——比较属性age的大小"

    print p1.__class__.__name__ + "##对应于java中的getclass,返回类名"
